util/goroutines/pool.go

```diff
@@ -0,0 +1,192 @@
+//      After a worker has been idle for a configured time, it will stop
+type Pool interface {
+	// Queue submits a unit of work to the pool. It will return an error if the pool is shutdown
+	Queue(func()) error
+	// QueueWithTimeout submits a unit of work to the pool. It will return an error if the pool is shutdown or
+	// if the work cannot be submitted to the work queue before the given timeout elapses
+	QueueWithTimeout(func(), time.Duration) error
+	// Shutdown stops all workers as they finish work and prevents new work from being submitted to the queue
+	Shutdown()
+}
+
+// PoolConfig is used to configure a new Pool
+type PoolConfig struct {
+	// The size of the channel feeding the worker pool
+	QueueSize uint32
+	// The minimum number of goroutines
+	MinWorkers uint32
+	// The maximum number of workers
+	MaxWorkers uint32
+	// IdleTime how long a goroutine should be idle before exiting
+	IdleTime time.Duration
+	// Provides a way to join shutdown of the pool with other components.
+	// The pool also be shut down independently using the Shutdown method
+	CloseNotify <-chan struct{}
+	// Provides a way to specify what happens if a worker encounters a panic
+	// if no PanicHandler is provided, panics will not be caught
+	PanicHandler func(err interface{})
+}
+
+func (self *PoolConfig) Validate() error {
+	if self.MaxWorkers < 1 {
+		return fmt.Errorf("max workers must be at least 1")
+	}
+	if self.MinWorkers > self.MaxWorkers {
+		return fmt.Errorf("min workers must be less than or equal to max workers. min workers=%v, max workers=%v", self.MinWorkers, self.MaxWorkers)
+	}
+
+	return nil
+}
+
+func NewPool(config PoolConfig) (Pool, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+	result := &pool{
+		queue:               make(chan func(), int(config.QueueSize)),
+		minWorkers:          config.MinWorkers,
+		maxWorkers:          config.MaxWorkers,
+		maxIdle:             config.IdleTime,
+		externalCloseNotify: config.CloseNotify,
+		closeNotify:         make(chan struct{}),
+		panicHandler:        config.PanicHandler,
+	}
+	for result.count < result.minWorkers {
+		result.tryAddWorker()
+	}
+	return result, nil
+}
+
+type pool struct {
+	queue               chan func()
+	count               uint32
+	minWorkers          uint32
+	maxWorkers          uint32
+	maxIdle             time.Duration
+	stopped             concurrenz.AtomicBoolean
+	externalCloseNotify <-chan struct{}
+	closeNotify         chan struct{}
+	panicHandler        func(err interface{})
+}
+
+func (self *pool) Queue(work func()) error {
+	if self.getCount() == 0 {
+		self.tryAddWorker()
+	}
+
+	select {
+	case self.queue <- work:
+		return nil
+	case <-self.closeNotify:
+		return errors.New("pool stopped")
+	case <-self.externalCloseNotify:
+		return errors.New("pool stopped (via external close notify)")
+	}
+}
+
+func (self *pool) QueueWithTimeout(work func(), timeout time.Duration) error {
+	if atomic.LoadUint32(&self.count) == 0 {
```

Comment:
Not a big thing, but same as getCount in case that ever changes?

Reply:
Ah, missed one. Fixed and consolidated queue logic in a single method. Also fixed an issue where panics could leave the pool empty.